lib/utils/task_queue/worker_queue.dart

```diff
@@ -0,0 +1,69 @@
+import 'dart:async';
+import 'dart:collection';
+
+import 'package:fluffychat/utils/task_queue/task.dart';
+import 'package:matrix/matrix.dart';
+
+typedef OnTaskCompleted = void Function(String? taskId);
+
+abstract class WorkerQueue {
+  final Queue<Task> queue = Queue<Task>();
+  Completer? completer;
+
+  String get workerName;
+
+  Future addTask(Task task) {
+    queue.add(task);
+    Logs().i(
+      'WorkerQueue<$workerName>::addTask(): QUEUE_LENGTH: ${queue.length}',
+    );
+    return _processTask();
+  }
+
+  Future _processTask() async {
+    if (completer != null) {
+      return completer!.future;
+    }
+    completer = Completer();
+    if (queue.isNotEmpty) {
+      final firstTask = queue.removeFirst();
+      Logs().i('WorkerQueue<$workerName>::_processTask(): ${firstTask.id}');
+      firstTask
+          .execute()
+          .then(_handleTaskExecuteCompleted)
+          .catchError(_handleTaskExecuteError)
+          .whenComplete(() => firstTask.onTaskCompleted?.call());
+    } else {
+      completer?.complete();
+    }
+    return completer!.future;
+  }
+
+  void _handleTaskExecuteCompleted(dynamic value) {
+    Logs().i('WorkerQueue<$workerName>::_handleTaskExecuteCompleted(): $value');
+    completer?.complete(value);
+    _releaseCompleter();
+    if (queue.isNotEmpty) {
```

Comment:
IMO this can should be in a separated method

Reply:
i think it's clear enough, why do you need to separate more ?